### ISSUE TITLE
Crabble - Code Review of Contract Code

### ISSUE THREAD

**Issue by aj-agoric** [Agoric/Agoric#8465]
## What is the Problem Being Solved?
Task: Code review the contract code which will be deployed to chain.

Purpose: Audit to raise confidence that contract implements desired behavior, is upgradable, properly uses durability & virtualization, doesn’t have bugs like infinite loops.

Instructions: OpCo engineers to arrange review with partner via Slack.

## Description of the Design

 - [x] check [timer watcher promises](#issuecomment-1816827151)
 - [x] check [vstorage collection structure](#issuecomment-1816841074)
 - [x] check that [swingset tests are repeated without upgrades in between](https://github.com/Agoric/agoric-sdk/issues/8465#issuecomment-1816875870)
 - [x] check that sequence diagrams have corresponding tests
 - [x] check that [work is limited by message size](https://github.com/Agoric/agoric-sdk/wiki/Limit-work-by-message-size)

**Comment by dckc** [comment]
@warner notes timer watcher promises might reject due to upgrade of the timer vat and the contract should use the right utility to handle that.

**Comment by dckc** [comment]
@michaelfig notes a best practice to avoid lists (rental0 / rental1) at the same level as non-lists; this suggests  moving that list under something like `rentals`

**Comment by dckc** [comment]
note to self on a testing style I was trying to suggest... IOU elaboration
cc @warner 

```js
  const wd = await walletFactoryDriver.provideSmartWallet('agoric1open');
  ```
https://github.com/Agoric/agoric-sdk/blob/87dbbda8484c6fe3fe542eb847647fd1540c11e6/packages/boot/test/bootstrapTests/test-vaults-integration.ts

**Comment by dckc** [comment]
For contract functionality in ci, Anil showed some nice swingset tests with upgrade between steps.

@warner suggested repeating the tests without upgrades in between. Anil agreed.

**Comment by dckc** [comment]
@anilhelvaci I took some notes above from the discussion we just had.

cc @otoole-brendan

**Comment by anilhelvaci** [comment]
Thanks! @dckc 

BTW, for that problem we had in the demo, it's because I had run out of IST in both of those accounts. So not a bug I believe 👌

**Comment by otoole-brendan** [comment]
Note to review deployment test (as part of partner checklist) during Agoric Opco code review

**Comment by anilhelvaci** [comment]
Here's an issue we run into when trying to deploy on Xnet => https://github.com/CrabblePitch/crabbleProtocol/issues/111#issuecomment-1820643312

**Comment by 0xpatrickdev** [comment]
When reviewing the code, we identified areas for improvement and worked with the team to address them. The issues we deemed critical to raising confidence were:

- getCatalog (PublicFacet) does O(n) work in response to O(1) input [#120](https://github.com/CrabblePitch/crabbleProtocol/issues/120)
- vstorage: avoid lists at the same level as non-lists [#127](https://github.com/CrabblePitch/crabbleProtocol/issues/127)
- rentalConfig Offer arguments need to have stricter limits [#135](https://github.com/CrabblePitch/crabbleProtocol/issues/135)
- insufficient issuer and brand verification in makeBidHook and makeRentalHook [#136](https://github.com/CrabblePitch/crabbleProtocol/issues/136)
- Invitation Proposal Shapes need to include exact terms to be offer safe [#138](https://github.com/CrabblePitch/crabbleProtocol/issues/138)
- Possible interleaving hazards in bidsManager [#139](https://github.com/CrabblePitch/crabbleProtocol/issues/139)
- zcf seat reallocations are not atomic, use atomicRearrange instead of multiple atomicTransfers [#140](https://github.com/CrabblePitch/crabbleProtocol/issues/140)
- makeHandle logic (kindHandle / label) [#150](https://github.com/CrabblePitch/crabbleProtocol/issues/150)
- Wrong `hour` value in `RentingDurationUnitToSeconds` [#157](https://github.com/CrabblePitch/crabbleProtocol/issues/157)
- bespoke contract governor lacks replaceElectorate [#158](https://github.com/CrabblePitch/crabbleProtocol/issues/158)
- rental bids vstorage shape needs improvement [#162](https://github.com/CrabblePitch/crabbleProtocol/issues/162)


All of these were addressed as of [CrabblePitch/crabbleProtocol#122c775310](https://github.com/CrabblePitch/crabbleProtocol/commit/122c7753107f2874a631135332875c1618d4a93a).

**Comment by dckc** [comment]
> note to self on a testing style I was trying to suggest... IOU elaboration cc @warner

I drafted something:

 - https://github.com/Agoric/agoric-sdk/discussions/8683
